### PR TITLE
[1316] Break up and parallelise build/test step

### PR DESCRIPTION
**Pull request by darokel**
### Context

https://trello.com/c/fNIyGyT5/5688-set-up-parallelisation-of-ci-tests

### Changes proposed in this pull request

Took a couple of punches from the CI but have managed to get the build down to around 8-10 mins-ish (from 30ish). 

- Add parallel testing gem and configure the Github workflows to use it
- Split out lint into its own step to also be parallelised 
- Add database cleaner gem to properly clean out tests (we seem to have issues with stale data not being cleared out)

### Guidance to review

It has been thoroughly reviewed by the CI believe me.

### CODE REVIEW

**Comment by darokel** [comment]
[Set up parallelisation of CI tests](https://trello.com/c/fNIyGyT5/5688-set-up-parallelisation-of-ci-tests)

**Review by defong — COMMENTED** [review]
A few niggly comments

File: Gemfile [hunk 1]
```diff
@@ -197,6 +199,7 @@ group :test do
   gem "webmock"
 
   # DfE Digital forked rspec-retry detects flaky specs and reports them
+  gem "database_cleaner-active_record"
```
[the remark this answers is not in the document]
Comment (darokel) [reply Gemfile:202]: done

File: .github/workflows/build-and-deploy.yml [hunk 2]
```diff
@@ -154,6 +262,8 @@ jobs:
       run: |
         docker-compose up --no-build -d
         docker-compose exec -T web /bin/sh -c "./wait-for-command.sh -c 'nc -z db 5432' -s 0 -t 20"
+        docker-compose exec -T web /bin/sh -c 'bundle config --local disable_exec_load true'
+        docker-compose exec -T web /bin/sh -c 'bundle exec rake parallel:setup'
         docker-compose exec -T web /bin/sh -c "yarn add jest@29.3.1"
```
[the remark this answers is not in the document]
Comment (darokel) [reply .github/workflows/build-and-deploy.yml:267]: yes, it's needed for the precompile step following it 😞

File: app/components/form_components/autocomplete/script.js [hunk 3]
```diff
@@ -13,7 +13,7 @@ const suggestion = (value, options) => {
     const html = option.append ? `<span>${value}</span> ${option.append}` : `<span>${value}</span>`
     return option.hint ? `${html}<br>${option.hint}` : html
   } else {
-    return `<span>No results found</span>`
+    return '<span>No results found</span>'
```
[the remark this answers is not in the document]
Comment (darokel) [reply app/components/form_components/autocomplete/script.js:16]: we're using https://standardjs.com, which is supposed to do that without the need to configure anything 🤷

**Comment by darokel** [comment]
FYI this is currently hanging because of the updated test names:

Screenshot 2023-07-17 at 14 44 07

It's setup as a required step in the branch protection rules:

Screenshot 2023-07-17 at 14 51 12

But the Test step has a unique name based on the matrix so Rick mentioned this may need to be removed once this is ready to be merged in (or we include every matrix name in the list)

**Comment by defong** [comment]
> FYI this is currently hanging because of the updated test names:
> Screenshot 2023-07-17 at 14 44 07
> 
> It's setup as a required step in the branch protection rules:
> Screenshot 2023-07-17 at 14 51 12
> 
> But the Test step has a unique name based on the matrix so Rick mentioned this may need to be removed once this is ready to be merged in (or we include every matrix name in the list)


https://docs.github.com/en/actions/using-jobs/using-jobs-in-a-workflow#defining-prerequisite-jobs

Will something like this work?

```yml
jobs:
  lint:
  test:
  must_pass
    needs: [lint, test]
  
  
```

**Comment by darokel** [comment]
> > FYI this is currently hanging because of the updated test names:
> > Screenshot 2023-07-17 at 14 44 07
> > It's setup as a required step in the branch protection rules:
> > Screenshot 2023-07-17 at 14 51 12
> > But the Test step has a unique name based on the matrix so Rick mentioned this may need to be removed once this is ready to be merged in (or we include every matrix name in the list)
> 
> https://docs.github.com/en/actions/using-jobs/using-jobs-in-a-workflow#defining-prerequisite-jobs
> 
> Will something like this work?
> 
> ```yaml
> jobs:
>   lint:
>   test:
>   must_pass
>     needs: [lint, test]
>   
>   
> ```

No I don't think that's relevant because see my comment about the test name changes.

File: README.md [hunk 4]
```diff
@@ -87,6 +87,21 @@ bundle exec rake
...
+
+`bundle exec rails parallel:setup`
+
+To run the specs in parallel:
+`bundle exec rails parallel:spec`
+
+To drop the test databases:
+`bundle exec rails parallel:drop`
+
+You can see the full list of commands with: `bundle exec rails -T | grep parallel`
```
Comment (stevehook) [inline README.md:103]: :+1:

**Review by defong — APPROVED** [review]
I'm happy to get 20 mins back for my life for each and every PR and CI reruns

Nice one @darokel :muscle:

**Review by carlosmartinez — APPROVED** [review]
excellent work!

**Pull request merged**